[Team22][iOS] 2주차 - 2 : DetailView 구현, 단위테스트 작성, Detail JSON 데이터 요청 흐름 구현

junu0516 opened the pull request:
안녕하세요 team22 주방, Jed 입니다.

마지막 PR을 보냅니다! 
2주간의 피드백을 통해 많은 인사이트 제공해주셔서 정말 감사합니다 :-)

### 작업 내용
- [x] 디테일뷰 구현, 스크롤 뷰 적용
- [x] 메인 뷰의 컬렉션뷰 셀 터치 시, 디테일 뷰를 위한 json 데이터 요청하는 흐름 구현
- [x] ReusableView 사용을 통한 컬렉션뷰의 각 섹션별 헤더뷰 적용
- [x] 메인 뷰의 컬렉션 뷰의 커스텀 셀 UI 개선(음식명, 설명, 가격, 뱃지 상하 간격)
- [x] Ordering 객체에 대한 단위테스트 작성

### 비고
- 마감시간에 쫓겨서 급하게 DetailView를 만들다 보니, 해당 클래스에 과하게 많은 컴포넌트들이 몰려버렸습니다. 현재 해당 파일의 코드가 약 400줄이 넘는데 이 부분은 주말 통해서 적당한 컴포넌트 단위로 파일을 분리하려 합니다!
- POST 요청을 보내는 흐름을 구현하지 못해서 이 부분도 주말 통해서 완성해볼까 합니다.
- 이전에 언급해주신 __`네트워크 응답받은 데이터와 캐싱된 데이터 간의 상이함 판단`__ 에 대해서는 어떻게 구현할 지 고민중입니다.
- 메인 뷰 내 뱃지 생성 시 일부 데이터 없는 경우 각 라벨 별 사이에 공간이 발생하게 되는데 이 공간을 없애고 나머지 데이터만 표시되도록 하고 싶은데 그부분에 대한 구현이 어려운데 남은 시간 적용을 해보려고 합니다.
- 드래그 시 뱃지 내용이 초기화 되지 않고 추가적으로 더해지고 있는데 이부분에 대해 초기화하는 부분을 적용 해보고자 합니다.

junu0516 committed:
[#11] Refactor: DataCacheable 프로토콜에 캐싱된 데이터 개수를 리턴하는 프로퍼티 추가

- DataCache 구현체 내부에서는 이에 대한 연산 프로퍼티를 구현
- 추후 데이터 캐싱이 잘 처리되었는 지 테스트하기 위해 추가

junu0516 committed:
[#11] Refactor: NetworkHandlable, RepositoryApplicable 프로토콜 메소드의 파라미터 타입 변경

- Data 타입을 파라미터로 받는 클로저가 아닌, Result<Data,Error> 타입을 파라미터로 받는 클로저를 사용하도록 변경

junu0516 committed:
[#11] Refactor: 뷰컨트롤러에서 네트워크 계층까지의 데이터 요청 흐름 수정, HttpError 열거형 추가

- completionHandler의 타입을 Result<Data,Error> 타입을 파라미터로 받는 클로저 타입으로 변경
- 뷰컨트롤러에서 최초로 completionHandler 정의 시에는, Result<Data,Error>의 success, failure 여부에 따라 각기 다른 로직을 처리하도록 함
- 네트워크 오류 발생 시에는 NetworkHandler에서 로그를 남기고 끝내는 것이 아닌, 오류 데이터를 클로저를 통해 보내 처리하도록 수정(추후 테스트가 용이하도록 하기 위함)

junu0516 committed:
[#11] Feat: MockRepository 수정, MockNetworkHandler, MockDataCache 추가

- 단위 테스트에 사용할 Mock Obejct 생성
- MockRepository는 Ordering을 테스트하기 위해 사용
- MockNetworkHandler, MockDataCache는 Ordering 및 Repository를 테스트하기 위해 사용

junu0516 committed:
[#11] Test: Ordering, Repository 클래스의 단위테스트 작성 및 주석 추가

junu0516 committed:
[#9] Feat: Repository에 convenience init 추가 후 MainViewController에서 NetworkHandler, DataCache 인스턴스 생성하는 로직을 Repository 내부로 이동

junu0516 committed:
[#9] Chore: git rebase 후 충돌나는 부분 수정(DataCachable -> DataCachingManagable 이름 수정 적용)

junu0516 committed:
[#9] Refactor: MockDataCache 이름을 MockDataCachingManager로 변경

junu0516 committed:
[#9] Feat: DetailViewController 구현 및 추가

junu0516 committed:
[#9] Feat: 메인 뷰에서 컬렉션 뷰 셀 클릭 시 json 데이터 요청 및 응답 확인 후 DetailViewController로 넘어가도록 구현

- 임시로 디버깅을 통해 응답 데이터를 확인했지만, 추후 객체로 디코딩해서 전달하도록 수정하고 로깅은 제거해야 함

junu0516 committed:
[#9] Add: MainResponse, DetailResponse 추가

- 기존의 Response는 MainResponse로 변경해서 사용
- DetailResponse 추가

junu0516 committed:
[#9] Chore: git rebase 충돌 해결 후 DataCachingManable에 누락되있던 count 속성 정의 추가, 불필요한 print() 호출 제거

junu0516 committed:
[#9] Feat: DetailResponse 속성 타입 변경, FoodDetail 추가 및 구현

junu0516 committed:
[#9] Feat: Repository의 기존 requestData를 요청 목적에 따라 requestBinaryData, requestModelData 함수로 분리

- 이미지 파일과 같은 단순 바이너리 데이터 요청 시에는 requestBinaryData를 호출
- 응답받은 데이터를 역직렬화한 모델 요청 시에는 requestModelData 호출

junu0516 committed:
[#9] Feat: MainViewController의 컬렉션뷰에서 셀 터치 시에 해당되는 FoodDetail 요청 후 다음 화면으로 넘어가는 흐름 구현

- Repository로부터 모델 데이터를 응답받아서 DetailViewController의 생성자로 넘겨줌
- 기존에 음식 이름을 Food.alt 속성을 사용하던 것을 Food.title 속성을 사용하도록 수정

junu0516 committed:
[#9] Refactor: MockRepository 생성 시에 Mock 객체들을 생성자로 주입받아야 하기 때문에 기존의 private init을 외부에서 접근 가능하도록 재수정

junu0516 committed:
[#9] Test: MockNetworkHandler, MockRepository 수정 및 RepositoryTest에 모델 데이터 요청하는 테스트 메소드 추가

junu0516 committed:
[#12] Feat: FoodDetailView 구현 및 추가(추후 세부 Component 별로 파일 분리 필요)

- 시간관계상 우선 FoodDetailView 안에 필요한 모든 서브 뷰를 추가했음
- 추후 각 섹션별로 개별 커스텀 UIView 클래스로 분리해야 함

junu0516 committed:
[#12] Feat: DetailViewController로 넘어갈 시 DetailView 레이아웃 설정 및 필요한 FoodDetail 데이터 화면에 나타나도록 구현

junu0516 committed:
[#12] Refactor: NetworkHandler 내부에서 http로 된 url을 https로 변경해주는 메소드 추가

junu0516 committed:
[#9] Chore: OrderingTest 속성에 수정된 타입명 반영

LeeJoobang committed:
[#6] Feat: 컬렉션 뷰마다 헤더 뷰 내용 변경 표시

LeeJoobang committed:
[#10] mainView UI 개선 진행 :  stackView 음식명, 정보, 가격, Badge 상하 간격 좁히는 작업

junu0516 committed:
[#10] Chore: git rebase 후 충돌나서 코드 꼬인 부분 해결

ITzombietux commented on SideDish/SideDish/Common/DataCachingManagable.swift:
```diff
@@ -1,6 +1,8 @@
 import Foundation
 
 protocol DataCachingManagable {
+    var count: Int { get }
+    
```
가능하면 count도 메서드로 알려주는건 어떻게 생각하시나요?ㅎㅎ

junu0516 replied:
구현체에서 해당 속성을 연산 프로퍼티로 구현하면 동일하게 getter의 역할을 할 수 있지 않을까 생각했습니다 ㅎㅎ
혹시 따로 getter 메소드 만드는걸 말씀하신걸까요!??

ITzombietux commented on SideDish/SideDish/Common/Mock/MockNetworkHandler.swift:
```diff
@@ -0,0 +1,55 @@
+    
+    var sampleJSONData: Data {
+        Data(
+                """
+                {
+                  "hash": "H72C3",
+                  "data": {
+                    "top_image": "http://public.codesquad.kr/jk/storeapp/data/soup/28_ZIP_P_1003_T.jpg",
+                    "thumb_images": [
+                      "http://public.codesquad.kr/jk/storeapp/data/soup/28_ZIP_P_1003_T.jpg",
+                      "http://public.codesquad.kr/jk/storeapp/data/soup/28_ZIP_P_1003_S.jpg"
+                    ],
+                    "product_description": "김치찌개에는 역시 돼지고기",
+                    "point": "83원",
+                    "delivery_info": "서울 경기 새벽 배송 / 전국 택배 배송",
+                    "delivery_fee": "2,500원 (40,000원 이상 구매 시 무료)",
+                    "prices": ["9,300원", "8,370원"],
+                    "detail_section": [
+                      "http://public.codesquad.kr/jk/storeapp/data/soup/28_ZIP_P_1003_D1.jpg",
+                      "http://public.codesquad.kr/jk/storeapp/data/soup/28_ZIP_P_1003_D2.jpg",
+                      "http://public.codesquad.kr/jk/storeapp/data/soup/28_ZIP_P_1003_D3.jpg",
+                      "http://public.codesquad.kr/jk/storeapp/data/pakage_regular.jpg"
+                    ]
+                  }
+                }
+
+                """.utf8
```
헛! 이 친구는 더미데이터인가요?ㅎㅎ 더미데이터라면 별도의 파일로 분리해서 사용한다면 좀 더 가독성이 올라갈거 같아요 ㅎㅎ!

junu0516 replied:
넵 별도의 json 파일을 만들어서 Bundle로 읽어들이는 식으로 바꿔보면 될 것 같습니다.
다만, 한가지 궁금한 점이 있는데 해당 Mock 객체를 가지고 테스트를 진행할 시에, 테스트를 위한 메소드 내부에서 테스트를 위해 Bundle에 있는 json 파일을 읽어들여도 괜찮을까요?

뭔가 테스트 메소드는 최대한 가볍고 빨라야 한다고 봐서 직접 서버에 데이터를 요청하거나, 로컬 혹은 다른 디렉토리에 있는 특정 파일을 읽어들이는 등의 행위를 하면 안될 것 같아서 위와 같이 그냥 속성으로 뒀습니다!
 (처음 이런 식으로 단위테스트 작성하는 걸 고민하다보니 이런 부분까지 다 마음에 걸리네요 😅)

ITzombietux commented on SideDish/SideDish/Common/Mock/MockNetworkHandler.swift:
```diff
@@ -0,0 +1,55 @@
+    
+    func request(url: EndPoint, method: HttpMethod, contentType: ContentType, completionHandler: @escaping (Result<Data,Error>) -> Void) {
+        
+        if !isResponseFailed{
+            switch contentType {
+            case .json:
+                completionHandler(.success(sampleJSONData))
+                delegate?.cachingDataRequested(url: url, data: sampleJSONData)
+            case .image:
+                let mockImageData = Data()
+                completionHandler(.success(mockImageData))
+                delegate?.cachingDataRequested(url: url, data: mockImageData)
+            }
+        }else{
+            completionHandler(.failure(HttpError.unknownError))
+        }
+    }
```
클린코드를 작성하려는 노력 중에 else를 사용하지 않는 것도 있어요! ㅎㅎ

junu0516 replied:
Mock 객체 만드는게 익숙하지 않다 보니 많이 클린하지 못한 것 같습..니다 ㅎㅎㅎ..
개인적으로 좀 더 공부해서 테스트하기 깔끔한 구조에 대해 고민해봐야겠습니다!

ITzombietux commented on SideDish/SideDish/Common/NetworkHandlable.swift:
```diff
@@ -2,5 +2,5 @@ import Foundation
 protocol NetworkHandlable{
     var delegate: NetworkHandlerDelegate? { get set }
-    func request(url: EndPoint, method: HttpMethod, contentType: ContentType, completionHandler: @escaping (Data) -> Void)
+    func request(url: EndPoint, method: HttpMethod, contentType: ContentType, completionHandler: @escaping (Result<Data,Error>)->Void)
```
Result타입을 사용한 것은 좋습니다!

ITzombietux commented on SideDish/SideDish/Controllers/DetailViewController.swift:
```diff
@@ -0,0 +1,75 @@
+
+final class DetailViewController: UIViewController{
+    
+    private var foodDetail: FoodDetail
+    private var foodTitle: String = ""
+    private let foodDetailView: FoodDetailView
+    weak var ordering: Ordering?
```
방학 기간에 MVVM을 공부한다고 하셨으니 이런 부분도 뷰컨트롤러와 분리하는 연습을 해보시면 좋을거 같습니다~!

ITzombietux commented on SideDish/SideDish/Controllers/DetailViewController.swift:
```diff
@@ -0,0 +1,75 @@
+    
+    //일정 상 썸네일 이미지는 스크롤뷰를 적용하기에는 학습시간이 부족해서 우선 이미지 한장만 적용
+    private func updateFoodThumbnailImage() {
+        ordering?.requesetFoodImage(imageUrl: foodDetail.thumbImageUrls[0]) { result in
```
thumbImageUrls[0] 으로 쓰기보다는 좀 더 의미있는 이름으로 사용한다면 가독성이 높아질거 같아요!

ITzombietux commented on SideDish/SideDish/Controllers/DetailViewController.swift:
```diff
@@ -0,0 +1,75 @@
+    
+    private func updateFoodDetailImages() {
+        foodDetailView.adjustFoodDetailImageStackViewLayout(imageCount: foodDetail.detailImageUrls.count)
+        for imageUrl in foodDetail.detailImageUrls {
+            ordering?.requesetFoodImage(imageUrl: imageUrl) { result in
+                switch result {
+                case .success(let imageData):
+                    DispatchQueue.main.async {
+                        self.foodDetailView.addDetailImage(imageData: imageData)
+                    }
+                case .failure(let error):
+                    self.logger.error("\(error.localizedDescription)")
+                }
+            }
```
이렇게 들여쓰기가 3~4단계까지 간다면 메서드를 분리할 수 있다는 의미가 돼요! ㅎㅎ

junu0516 replied:
넵 다시 보내 해당 컨트롤러에서 저런 식으로 Result 타입을 가지고 비슷한 양식으로 switch문을 작성하는게 곳곳에 있는데 공통 메소드를 하나 작성해야겠습니다!

ITzombietux commented on SideDish/SideDish/Controllers/MainViewController.swift:
```diff
@@ -1,25 +1,37 @@
 class MainViewController: UIViewController {
     
     private var ordering: Ordering?
+    private var logger: Logger?
```
DetailViewController에서는 logger가 옵셔널이 아니고 여기선 옵셔널인데 의도가 있을까요??ㅎㅎ

junu0516 replied:
음 굳이 옵셔널일 필요가 없는데 옵셔널을 사용한 것 같습니다! 다시보니 DetailViewController와 동일하게 옵셔널이 아닌 식으로 적용하는게 나아 보입니다!

ITzombietux commented on SideDish/SideDish/Controllers/MainViewController.swift:
```diff
@@ -59,14 +78,48 @@ extension MainViewController: UICollectionViewDelegate, UICollectionViewDataSour
+        guard let food = ordering[indexPath.row,category] else { return }
+        
+        ordering.requestFoodDetail(detailHash: food.detailHash){ result in
+            switch result {
+            case .success(let response):
+                self.moveToDetailView(foodDetail: response.data, foodTitle: food.title)
+            case .failure(let error):
+                self.logger?.error("\(error.localizedDescription)")
+            }
```
이런 로직도 MVVM에서 어떻게 동작하는지 알아보는 것도 좋을거 같아요~!

ITzombietux reviewed (commented):
프로젝트 진행하느라 대단히 수고많으셨습니다.!

방학 동안 이 프로젝트를 진행하신다면 
- MVVM을 공부하시면서 개체 마다 명확하게 역할 분리
- 메서드는 하나의 역할만!
- else 사용해보지 않기 (2가지 경우가 아닌 늘어날 가능성이 있는 로직)
- 들여쓰기 2개 이상 하지 않기
이런 연습들도 해보시는걸 추천합니다!! 하시면서 다시 PR보내시면 그때 또 한번 소통했으면 좋겠어요 ㅎㅎ

junu0516 commented on SideDish/SideDish/Controllers/DetailViewController.swift:
```diff
@@ -1,13 +1,18 @@
     private var foodDetail: FoodDetail
     private var foodTitle: String = ""
+    private let foodDetailView: FoodDetailView
+    weak var ordering: Ordering?
```
MainViewController에서 NavigationViewController을 통해서 DetailViewController로 넘어가면서 MainviewController가 기존에 참조하고 있던 Ordering 인스턴스를 약한 참조 형태로 DetailViewController 인스턴스에 주입시켰습니다!

이런 방법을 사용한 이유는 DetailViewController에서 주문할 음식 수량 선택 혹은 주문하기 같은 부분을 위해 Ordering으로 특정 로직 수행을 요청해야 되기 때문에 동일한 Ordering 인스턴스를 참조해야 한다고 생각했기 때문입니다.

혹시 이런 식으로 동일한 모델 인스턴스를 참조하게 하는 방식이 괜찮은건지 궁금합니다!

ITzombietux replied:
네 전혀 문제가 없습니다. ㅎㅎㅎ  하지만 

```.swift
detailViewController.ordering = ordering
```

이런 식으로 속성에 직접 접근하기 보다는 메서드 / 생성자 를 사용하여 의존성 주입을 시키면 더 좋아요 ㅎㅎ

junu0516 commented on SideDish/SideDish/Controllers/MainViewController.swift:
```diff
@@ -48,6 +48,7 @@ class MainViewController: UIViewController {
     
     func moveToDetailView(foodDetail: FoodDetail, foodTitle: String) {
         let detailViewController = DetailViewController(foodDetail: foodDetail, foodTitle: foodTitle)
+        detailViewController.ordering = ordering
```
해당 부분이 DetailViewController 인스턴스 생성 후 MainViewController가 참조하고 있던 Ordering 인스턴스를 약한 참조로 주입시키는 부분입니다!

junu0516 committed:
Chore: MainViewController.logger 옵셔널 타입이던 부분 수정

- 굳이 옵셔널일 필요가 없기 때문에 필드에서 Logger 인스턴스를 아얘 초기화

junu0516 committed:
Feat: MainViewController의 셀 터치 시 DetailView로 넘어가는 흐름을 옵저버 패턴을 적용하여 수정

- 셀 터치 시 바로 Ordering.selectFood를 호출하는 식으로 수정
- Ordering 내부에서 Repository로 데이터를 요청하는 메소드 호출
- 정상적으로 응답 수신 시에 NotificationCenter을 이용하여 다시 뷰컨트롤러 - 뷰 의 방향으로 화면 출력을 요청하는 흐름을 적용

junu0516 committed:
Add: FoodDetailViewDelegate 프로토콜 추가

junu0516 committed:
Refactor: DetailViewController의 생성자로 Ordering 인스턴스를 주입받아서 약한 참조하도록 수정, 화면에 보여줄 FoodDetail을 Ordering.selectedFoodDetail 속성을 읽어들이도록 수정

junu0516 committed:
Feat: FoodDetailView의 UIStepper 요소 터치 시, 선택된 음식의 주문수량 변경 및 주문 총액 수정하는 기능 추가

- FoodDetailView -> DetailViewController로의 흐름은 Delegate 패턴 적용
- Ordering -> DetailViewController 로의 흐름은 NotificationCenter을 사용한 옵저버 패턴 적용
- Stepper의 값을 그대로 전달해서 사용하기 때문에 Ordering 내부의 increaseCount 와 같은 메소드를 제거하고 setOrderingCount 메소드를 사용하여 전달받은 값에 맞춰 수량과 주문총액을 바로 연산하도록 수정

junu0516 committed:
Chore: 코드 들여쓰기 정리

junu0516 committed:
Chore: 디테일 형식 정리
